### PR TITLE
Dependency injection

### PR DESCRIPTION
**Pull request by jcorteza**

### CODE REVIEW

File: src/main/java/com/khoros/twitterapp/TwitterApp.java [hunk 1]
```diff
@@ -3,20 +3,31 @@
...
+    @Component(modules = {ServiceProviderModule.class})
+    public interface ResourceComponent {
+        MainResource mainResource();
+    }
+
     @Override
     public void run(TwitterAppConfiguration configuration, Environment environment) {
 
         TwitterService.getInstance().setTWFactory(configuration.twitter4jConfigurationBuild().build());
+        ResourceComponent resourceComponent = DaggerTwitterApp_ResourceComponent.builder()
```
Comment (kev-chan) [inline src/main/java/com/khoros/twitterapp/TwitterApp.java]: in the line above, if you are properly using dependency injection, you should not need to "setTWFactory" explicit anywhere.

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/main/java/com/khoros/twitterapp/TwitterApp.java (hunk 1 above)
Comment (kev-chan) [reply src/main/java/com/khoros/twitterapp/TwitterApp.java]: look into injecting everything you need

File: src/main/java/com/khoros/twitterapp/TwitterApp.java [hunk 2]
```diff
@@ -1,22 +1,49 @@
...
     public void run(TwitterAppConfiguration configuration, Environment environment) {
 
-        TwitterService.getInstance().setTWFactory(configuration.twitter4jConfigurationBuild().build());
+        Configuration twConfig = configuration.twitter4jConfigurationBuild().build();
+        Twitter twFactory = new TwitterFactory(twConfig).getInstance();
+        TwitterServiceComponent twServiceComponent = DaggerTwitterApp_TwitterServiceComponent.builder()
+                .twitterFactoryModule(new TwitterFactoryModule(twFactory))
+                .build();
+        // use injectTwitterService method to injection Twitter object
+        twServiceComponent.injectTwitterService(TwitterService.getInstance());
```
Comment (kev-chan) [inline src/main/java/com/khoros/twitterapp/TwitterApp.java]: does this need to be here now that we have injection?

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/main/java/com/khoros/twitterapp/TwitterApp.java (hunk 2 above)
Comment (jcorteza) [reply src/main/java/com/khoros/twitterapp/TwitterApp.java]: the injectTwitterService method? I had to add it to my component for the field injection to work. This was a solution I found from a video I watched. It's not actually part of TwitterService code

File: src/main/java/com/khoros/twitterapp/TwitterApp.java [hunk 3]
```diff
@@ -1,22 +1,38 @@
...
+    public interface ResourceComponent {
+        MainResource mainResource();
+    }
+
     @Override
     public void run(TwitterAppConfiguration configuration, Environment environment) {
 
-        TwitterService.getInstance().setTWFactory(configuration.twitter4jConfigurationBuild().build());
+        Configuration twConfig = configuration.twitter4jConfigurationBuild().build();
+        Twitter twFactory = new TwitterFactory(twConfig).getInstance();
```
Comment (kev-chan) [inline src/main/java/com/khoros/twitterapp/TwitterApp.java]: same comment as before, use injection for this

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/main/java/com/khoros/twitterapp/services/TwitterService.java [hunk 4]
```diff
@@ -7,31 +7,25 @@
...
-    private static Twitter twitterFactory;
+    public Twitter twitterFactory;
 
-    private TwitterService() {
-        // hidden constructor
-    }
-
-    public static TwitterService getInstance() {
-        return INSTANCE;
+    @Inject public TwitterService(Twitter twitterFactory) {
```
Comment (kev-chan) [inline src/main/java/com/khoros/twitterapp/services/TwitterService.java]: put @Inject above the constructor, it makes it easier to read.  Nice job!

File: src/main/java/com/khoros/twitterapp/TwitterApp.java [hunk 5]
```diff
@@ -1,22 +1,48 @@
...
+import dagger.Component;
+import javax.inject.Singleton;
 
 public class TwitterApp extends Application<TwitterAppConfiguration> {
 
     public static void main(String[] args) throws Exception {
         new TwitterApp().run(args);
     }
 
+    @Component(modules = TwitterFactoryModule.class)
```
Comment (kev-chan) [inline src/main/java/com/khoros/twitterapp/TwitterApp.java]: make the component a separate class vs inlining it in the application

File: src/main/java/com/khoros/twitterapp/TwitterFactoryModule.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+@Module
+class TwitterFactoryModule {
+    private Configuration twConf;
+
+    TwitterFactoryModule(Configuration twConf) {
+        this.twConf = twConf;
+    }
+
+    @Provides Twitter provideTwitterFactory() {
```
Comment (kev-chan) [inline src/main/java/com/khoros/twitterapp/TwitterFactoryModule.java]: annotation on a separate line

File: src/test/java/com/khoros/twitterapp/TwitterServiceTest.java [hunk 7]
```diff
@@ -28,9 +24,9 @@
     @Before
     public void setup() {
 
-        twSingleton = TwitterService.getInstance();
         mockFactory = mock(Twitter.class);
-        twSingleton.setTWFactory(mockFactory);
+        twSingleton = new TwitterService(mockFactory);
+        // twSingleton.setTWFactory(mockFactory);
```
Comment (kev-chan) [inline src/test/java/com/khoros/twitterapp/TwitterServiceTest.java]: you can delete this if it isn't needed

**Review by kev-chan — CHANGES_REQUESTED** [review]

File: src/test/java/com/khoros/twitterapp/TwitterServiceTest.java [hunk 8]
```diff
@@ -3,22 +3,18 @@
 import com.khoros.twitterapp.services.TwitterService;
 import com.khoros.twitterapp.models.Status;
 
-import static org.mockito.Mockito.*;
-
 import com.khoros.twitterapp.services.TwitterServiceException;
+
+import static org.mockito.Mockito.*;
```
Comment (kev-chan) [inline src/test/java/com/khoros/twitterapp/TwitterServiceTest.java]: don't use *

File: src/test/java/com/khoros/twitterapp/MainResourceTest.java [hunk 9]
```diff
@@ -8,8 +8,8 @@
 
 import static org.mockito.Mockito.*;
```
Comment (kev-chan) [inline src/test/java/com/khoros/twitterapp/MainResourceTest.java]: same

File: src/main/java/com/khoros/twitterapp/TwitterApp.java [hunk 10]
```diff
@@ -15,8 +12,18 @@ public static void main(String[] args) throws Exception {
     @Override
     public void run(TwitterAppConfiguration configuration, Environment environment) {
 
-        TwitterService.getInstance().setTWFactory(configuration.twitter4jConfigurationBuild().build());
+        TwitterFactoryComponent twComponent = DaggerTwitterFactoryComponent.builder()
+                .twitterFactoryModule(
+                        new TwitterFactoryModule(
+                                // Twitter Configuration Object
+                                configuration.twitter4jConfigurationBuild().build()
```
Comment (kev-chan) [inline src/main/java/com/khoros/twitterapp/TwitterApp.java]: better implementation, leave the configuration as just getters and setters and move all of the building to the provider

**Review by kev-chan — CHANGES_REQUESTED** [review]
fix the * in mockito and the twitter config. You can merge once that is done

**Pull request merged**